File: admin/src/utils/brandPalette.ts
/**
 * Copy of `client/utils/brandPalette.ts` for the admin panel's preview.
 *
 * The admin is built from its own directory (`COPY admin/ .` in docker/admin.Dockerfile),
 * so it cannot import across the repository. The two files are kept identical below this
 * header; `client/utils/brandPalette.shared.test.ts` fails when they drift. Change the
 * client file first, then copy it here.
 */
/**
 * Turns the one colour an operator picks into everything a theme needs from it: the
 * eleven-step scale Tailwind utilities expect, and the text colour that stays readable on
 * top of it.
 *
 * Kept free of Nuxt so the storefront (server-side, into `:root`), the admin preview and
 * the e-mail renderer all agree. The e-mail side is `BrandPalette.cs` on the backend, a
 * port of this file; the two carry the same test numbers and are changed together.
 *
 * The ramp is built in OKLCH: lightness is spread from near-white (50) to near-black
 * (950) while hue stays put and chroma tapers toward the ends, which is what keeps a
 * yellow from turning brown at 800 and a blue from going grey at 200. The picked colour
 * is pinned at 500, so `bg-primary-500` is exactly what the operator chose.
 *
 * @module utils/brandPalette
 */

/** One shade as sRGB 0–255 integers. */
export type Rgb = readonly [number, number, number]

/** The Tailwind shade steps, in order. */
export const SHADE_STEPS = [50, 100, 200, 300, 400, 500, 600, 700, 800, 900, 950] as const

/** A shade step. */
export type ShadeStep = (typeof SHADE_STEPS)[number]

/** A full scale keyed by step. */
export type Scale = Record<ShadeStep, Rgb>

/**
 * OKLCH lightness for each step. 500 is replaced by the picked colour's own lightness,
 * and the steps either side are re-spaced so the ramp stays monotonic whatever was picked.
 */
const TARGET_L: Record<ShadeStep, number> = {
  50: 0.975, 100: 0.945, 200: 0.89, 300: 0.81, 400: 0.72,
  500: 0.62, 600: 0.53, 700: 0.45, 800: 0.38, 900: 0.31, 950: 0.22,
}

/**
 * Parses `#rrggbb` (any case) into sRGB integers.
 *
 * @param hex The colour, with its `#`.
 * @returns The channels, or `null` when the string is not a six-digit hex colour.
 */
export function hexToRgb(hex: string): Rgb | null {
  const m = /^#([0-9a-f]{6})$/i.exec(hex.trim())
  if (!m) return null
  const n = parseInt(m[1]!, 16)
  return [(n >> 16) & 255, (n >> 8) & 255, n & 255]
}

/**
 * Formats sRGB integers as lower-case `#rrggbb`.
 *
 * @param rgb The channels.
 * @returns The hex string.
 */
export function rgbToHex([r, g, b]: Rgb): string {
  return '#' + [r, g, b].map(c => c.toString(16).padStart(2, '0')).join('')
}

/** sRGB 0–255 → linear 0–1. */
const toLinear = (c: number): number => {
  const s = c / 255
  return s <= 0.04045 ? s / 12.92 : ((s + 0.055) / 1.055) ** 2.4
}

/** Linear 0–1 → sRGB 0–255, clamped. */
const toSrgb = (l: number): number => {
  const c = l <= 0.0031308 ? l * 12.92 : 1.055 * l ** (1 / 2.4) - 0.055
  return Math.round(Math.min(1, Math.max(0, c)) * 255)
}

/** sRGB → OKLab (Björn Ottosson's matrices). */
function rgbToOklab([r, g, b]: Rgb): [number, number, number] {
  const lr = toLinear(r), lg = toLinear(g), lb = toLinear(b)
  const l = Math.cbrt(0.4122214708 * lr + 0.5363325363 * lg + 0.0514459929 * lb)
  const m = Math.cbrt(0.2119034982 * lr + 0.6806995451 * lg + 0.1073969566 * lb)
  const s = Math.cbrt(0.0883024619 * lr + 0.2817188376 * lg + 0.6299787005 * lb)
  return [
    0.2104542553 * l + 0.7936177850 * m - 0.0040720468 * s,
    1.9779984951 * l - 2.4285922050 * m + 0.4505937099 * s,
    0.0259040371 * l + 0.7827717662 * m - 0.8086757660 * s,
  ]
}

/** OKLab → linear sRGB, unclamped, so a caller can tell whether the colour is in gamut. */
function oklabToLinear([L, a, b]: [number, number, number]): [number, number, number] {
  const l = (L + 0.3963377774 * a + 0.2158037573 * b) ** 3
  const m = (L - 0.1055613458 * a - 0.0638541728 * b) ** 3
  const s = (L - 0.0894841775 * a - 1.2914855480 * b) ** 3
  return [
    +4.0767416621 * l - 3.3077115913 * m + 0.2309699292 * s,
    -1.2684380046 * l + 2.6097574011 * m - 0.3413193965 * s,
    -0.0041960863 * l - 0.7034186147 * m + 1.7076147010 * s,
  ]
}

/** Whether every linear channel is displayable, with a hair of tolerance for rounding. */
const inGamut = (lin: [number, number, number]): boolean =>
  lin.every(c => c >= -0.0005 && c <= 1.0005)

/**
 * OKLCH → sRGB, mapped into gamut by reducing chroma at fixed lightness and hue.
 *
 * Clamping channels instead — what a naive conversion does — turns a light tint of a
 * saturated blue into cyan and a light yellow into acid green, because the clipped channel
 * changes the hue. Pulling chroma back until the colour fits keeps the hue the operator
 * chose; the tint just gets a little paler, which is what a tint is.
 */
function oklchToRgb(L: number, chroma: number, hue: number): Rgb {
  const at = (c: number): [number, number, number] =>
    oklabToLinear([L, c * Math.cos(hue), c * Math.sin(hue)])

  let lin = at(chroma)
  if (!inGamut(lin)) {
    let lo = 0, hi = chroma
    for (let i = 0; i < 20; i++) {
      const mid = (lo + hi) / 2
      if (inGamut(at(mid))) lo = mid
      else hi = mid
    }
    lin = at(lo)
  }
  return [toSrgb(lin[0]), toSrgb(lin[1]), toSrgb(lin[2])]
}

/**
 * Builds the eleven-step scale around one colour.
 *
 * @param hex The picked colour, `#rrggbb`.
 * @returns The scale, with the picked colour itself at 500.
 * @throws {RangeError} When `hex` is not a six-digit hex colour.
 */
export function buildScale(hex: string): Scale {
  const rgb = hexToRgb(hex)
  if (!rgb) throw new RangeError(`Not a #rrggbb colour: ${hex}`)

  const [L0, a0, b0] = rgbToOklab(rgb)
  const chroma0 = Math.hypot(a0, b0)
  const hue = Math.atan2(b0, a0)

  const scale = {} as Scale
  for (const step of SHADE_STEPS) {
    if (step === 500) {
      scale[step] = rgb
      continue
    }
    // Re-space the target lightness so the picked colour sits at 500 and the ramp stays
    // monotonic: steps above 500 interpolate between L0 and white, steps below between
    // L0 and black, each keeping its relative position from the reference table.
    // The ends move with the picked colour: a near-black brand still gets a darker 950 and a
    // near-white one a lighter 50, so the ramp never flattens into equal steps.
    const top = Math.max(0.985, L0 + (1 - L0) * 0.6)
    const bottom = Math.min(0.15, L0 * 0.6)
    const ref = TARGET_L[step]
    const ref500 = TARGET_L[500]
    const L = step < 500
      ? L0 + (ref - ref500) / (TARGET_L[50] - ref500) * (top - L0)
      : L0 - (ref500 - ref) / (ref500 - TARGET_L[950]) * (L0 - bottom)
    // Chroma tapers toward both ends so the light steps read as tints rather than pastel
    // neon and the dark ones do not go muddy. Tints taper harder (to ~8% at 50) than shades
    // (to ~25% at 950): a saturated near-white is out of sRGB gamut and clamps to cyan or
    // acid yellow, while a dark shade keeps its identity with more colour left in it.
    const distance = Math.min(1, Math.abs(L - L0) / Math.max(L0 - bottom, top - L0))
    const taper = step < 500 ? 0.92 : 0.75
    const chroma = chroma0 * (1 - taper * distance ** 0.8)
    scale[step] = oklchToRgb(L, chroma, hue)
  }
  return scale
}

/**
 * WCAG 2 relative luminance.
 *
 * @param rgb The channels.
 * @returns Luminance in 0–1.
 */
export function luminance([r, g, b]: Rgb): number {
  return 0.2126 * toLinear(r) + 0.7152 * toLinear(g) + 0.0722 * toLinear(b)
}

/**
 * WCAG 2 contrast ratio between two colours, 1–21.
 *
 * @param a One colour, `#rrggbb`.
 * @param b The other, `#rrggbb`.
 * @returns The ratio, or `NaN` when either string is not a colour.
 */
export function contrastRatio(a: string, b: string): number {
  const ra = hexToRgb(a), rb = hexToRgb(b)
  if (!ra || !rb) return NaN
  const la = luminance(ra), lb = luminance(rb)
  const [hi, lo] = la > lb ? [la, lb] : [lb, la]
  return (hi + 0.05) / (lo + 0.05)
}

/** The AA threshold for normal text. */
export const WCAG_AA = 4.5

/** The AA threshold for large or bold text, and the point below which white stops being an option. */
export const WCAG_AA_LARGE = 3

/**
 * The text colour to put on a filled surface of the given colour.
 *
 * White wins whenever it reaches the large-text threshold (3:1) — buttons are bold, and
 * white on a mid blue or purple is what every design system ships even where dark text
 * would score a little higher. Below that, whichever of white and near-black reads better:
 * a brand yellow or a pale teal gets dark text.
 *
 * @param hex The surface colour, `#rrggbb`.
 * @returns `#ffffff` or `#111111`.
 */
export function bestTextOn(hex: string): '#ffffff' | '#111111' {
  const white = contrastRatio('#ffffff', hex)
  if (white >= WCAG_AA_LARGE) return '#ffffff'
  return white >= contrastRatio('#111111', hex) ? '#ffffff' : '#111111'
}

/**
 * Renders a scale as the `--<name>-<step>: r g b` declarations the stylesheets read, in
 * the space-separated form Tailwind's `<alpha-value>` slot needs.
 *
 * @param name The variable stem, e.g. `brand-primary`.
 * @param scale The scale.
 * @returns One declaration per step, joined with `;`.
 */
export function scaleToDeclarations(name: string, scale: Scale): string {
  return SHADE_STEPS.map(step => `--${name}-${step}:${scale[step].join(' ')}`).join(';')
}

File: backend/src/Innovayse.Application/Notifications/Services/BrandPalette.cs
namespace Innovayse.Application.Notifications.Services;

using System.Globalization;

/// <summary>
/// Derives the shades and the readable text colour for the brand colour an operator picked,
/// for the e-mail renderer.
/// </summary>
/// <remarks>
/// <para>
/// A port of <c>client/utils/brandPalette.ts</c>, which is the authority: the storefront
/// emits its <c>:root</c> variables and the admin panel draws its preview from that file, and
/// an e-mail must use the same numbers or a button in the inbox will not match the button on
/// the site. <c>BrandPaletteTests</c> carries the same expected values as the TypeScript test
/// so that a change to one side without the other fails a build.
/// </para>
/// <para>
/// The ramp is built in OKLCH — lightness spread from near-white (50) to near-black (950),
/// hue fixed, chroma tapered toward the ends and pulled back into the sRGB gamut rather than
/// clamped — so a tint of a saturated blue stays blue instead of turning cyan. The picked
/// colour itself sits at 500.
/// </para>
/// </remarks>
public static class BrandPalette
{
    /// <summary>The shade steps, in Tailwind's order.</summary>
    public static readonly IReadOnlyList<int> Steps = [50, 100, 200, 300, 400, 500, 600, 700, 800, 900, 950];

    /// <summary>White, the first candidate for text on a filled surface.</summary>
    public const string White = "#ffffff";

    /// <summary>Near-black, the text colour for a surface too light for white.</summary>
    public const string Ink = "#111111";

    /// <summary>Contrast ratio at which white is accepted on a filled surface (WCAG AA, large text).</summary>
    private const double WhiteThreshold = 3.0;

    /// <summary>Reference OKLCH lightness per step; 500 is replaced by the picked colour's own.</summary>
    private static readonly IReadOnlyDictionary<int, double> TargetL = new Dictionary<int, double>
    {
        [50] = 0.975, [100] = 0.945, [200] = 0.89, [300] = 0.81, [400] = 0.72,
        [500] = 0.62, [600] = 0.53, [700] = 0.45, [800] = 0.38, [900] = 0.31, [950] = 0.22,
    };

    /// <summary>
    /// Parses <c>#rrggbb</c> in any case.
    /// </summary>
    /// <param name="hex">The colour with its <c>#</c>.</param>
    /// <returns>The channels 0–255, or <see langword="null"/> when the string is not a colour.</returns>
    public static (int R, int G, int B)? Parse(string hex)
    {
        var s = hex.Trim();
        if (s.Length != 7 || s[0] != '#'
            || !int.TryParse(s.AsSpan(1), NumberStyles.HexNumber, CultureInfo.InvariantCulture, out var n))
        {
            return null;
        }

        return ((n >> 16) & 255, (n >> 8) & 255, n & 255);
    }

    /// <summary>
    /// Formats channels as lower-case <c>#rrggbb</c>.
    /// </summary>
    /// <param name="rgb">The channels.</param>
    /// <returns>The hex string.</returns>
    public static string Format((int R, int G, int B) rgb) =>
        string.Create(CultureInfo.InvariantCulture, $"#{rgb.R:x2}{rgb.G:x2}{rgb.B:x2}");

    /// <summary>
    /// One shade of the scale built around a colour.
    /// </summary>
    /// <param name="hex">The picked colour, <c>#rrggbb</c>.</param>
    /// <param name="step">A member of <see cref="Steps"/>.</param>
    /// <returns>The shade as <c>#rrggbb</c>; the input itself for step 500.</returns>
    /// <exception cref="ArgumentException">Not a colour, or not a step.</exception>
    public static string Shade(string hex, int step)
    {
        var rgb = Parse(hex) ?? throw new ArgumentException($"Not a #rrggbb colour: {hex}", nameof(hex));
        if (!TargetL.TryGetValue(step, out var refL))
        {
            throw new ArgumentException($"Not a shade step: {step}", nameof(step));
        }

        if (step == 500)
        {
            return Format(rgb);
        }

        var (l0, a0, b0) = RgbToOklab(rgb);
        var chroma0 = Math.Sqrt(a0 * a0 + b0 * b0);
        var hue = Math.Atan2(b0, a0);

        // Same construction as the TypeScript: the ends move with the picked colour so the
        // ramp never flattens, and tints taper chroma harder than shades.
        var top = Math.Max(0.985, l0 + (1 - l0) * 0.6);
        var bottom = Math.Min(0.15, l0 * 0.6);
        var ref500 = TargetL[500];
        var l = step < 500
            ? l0 + (refL - ref500) / (TargetL[50] - ref500) * (top - l0)
            : l0 - (ref500 - refL) / (ref500 - TargetL[950]) * (l0 - bottom);
        var distance = Math.Min(1, Math.Abs(l - l0) / Math.Max(l0 - bottom, top - l0));
        var taper = step < 500 ? 0.92 : 0.75;
        var chroma = chroma0 * (1 - taper * Math.Pow(distance, 0.8));

        return Format(OklchToRgb(l, chroma, hue));
    }

    /// <summary>
    /// WCAG 2 contrast ratio between two colours, 1–21.
    /// </summary>
    /// <param name="a">One colour.</param>
    /// <param name="b">The other.</param>
    /// <returns>The ratio, or <see cref="double.NaN"/> when either is not a colour.</returns>
    public static double ContrastRatio(string a, string b)
    {
        if (Parse(a) is not { } ra || Parse(b) is not { } rb)
        {
            return double.NaN;
        }

        var la = Luminance(ra);
        var lb = Luminance(rb);
        var (hi, lo) = la > lb ? (la, lb) : (lb, la);
        return (hi + 0.05) / (lo + 0.05);
    }

    /// <summary>
    /// The text colour for a filled surface: white when it reaches 3:1, otherwise whichever
    /// of white and near-black reads better. Identical to the TypeScript rule.
    /// </summary>
    /// <param name="hex">The surface colour.</param>
    /// <returns><see cref="White"/> or <see cref="Ink"/>.</returns>
    public static string BestTextOn(string hex)
    {
        var white = ContrastRatio(White, hex);
        if (white >= WhiteThreshold)
        {
            return White;
        }

        return white >= ContrastRatio(Ink, hex) ? White : Ink;
    }

    /// <summary>WCAG 2 relative luminance.</summary>
    private static double Luminance((int R, int G, int B) rgb) =>
        0.2126 * ToLinear(rgb.R) + 0.7152 * ToLinear(rgb.G) + 0.0722 * ToLinear(rgb.B);

    /// <summary>sRGB 0–255 → linear 0–1.</summary>
    private static double ToLinear(int c)
    {
        var s = c / 255.0;
        return s <= 0.04045 ? s / 12.92 : Math.Pow((s + 0.055) / 1.055, 2.4);
    }

    /// <summary>Linear 0–1 → sRGB 0–255, clamped.</summary>
    private static int ToSrgb(double l)
    {
        var c = l <= 0.0031308 ? l * 12.92 : 1.055 * Math.Pow(l, 1 / 2.4) - 0.055;
        return (int)Math.Round(Math.Clamp(c, 0, 1) * 255, MidpointRounding.AwayFromZero);
    }

    /// <summary>sRGB → OKLab.</summary>
    private static (double L, double A, double B) RgbToOklab((int R, int G, int B) rgb)
    {
        var lr = ToLinear(rgb.R);
        var lg = ToLinear(rgb.G);
        var lb = ToLinear(rgb.B);
        var l = Math.Cbrt(0.4122214708 * lr + 0.5363325363 * lg + 0.0514459929 * lb);
        var m = Math.Cbrt(0.2119034982 * lr + 0.6806995451 * lg + 0.1073969566 * lb);
        var s = Math.Cbrt(0.0883024619 * lr + 0.2817188376 * lg + 0.6299787005 * lb);
        return (
            0.2104542553 * l + 0.7936177850 * m - 0.0040720468 * s,
            1.9779984951 * l - 2.4285922050 * m + 0.4505937099 * s,
            0.0259040371 * l + 0.7827717662 * m - 0.8086757660 * s);
    }

    /// <summary>OKLab → linear sRGB, unclamped.</summary>
    private static (double R, double G, double B) OklabToLinear(double L, double a, double b)
    {
        var l = Math.Pow(L + 0.3963377774 * a + 0.2158037573 * b, 3);
        var m = Math.Pow(L - 0.1055613458 * a - 0.0638541728 * b, 3);
        var s = Math.Pow(L - 0.0894841775 * a - 1.2914855480 * b, 3);
        return (
            +4.0767416621 * l - 3.3077115913 * m + 0.2309699292 * s,
            -1.2684380046 * l + 2.6097574011 * m - 0.3413193965 * s,
            -0.0041960863 * l - 0.7034186147 * m + 1.7076147010 * s);
    }

    /// <summary>Whether every linear channel is displayable.</summary>
    private static bool InGamut((double R, double G, double B) lin) =>
        lin.R is >= -0.0005 and <= 1.0005 && lin.G is >= -0.0005 and <= 1.0005 && lin.B is >= -0.0005 and <= 1.0005;

    /// <summary>OKLCH → sRGB, reducing chroma at fixed lightness and hue until the colour fits.</summary>
    private static (int R, int G, int B) OklchToRgb(double l, double chroma, double hue)
    {
        (double, double, double) At(double c) => OklabToLinear(l, c * Math.Cos(hue), c * Math.Sin(hue));

        var lin = At(chroma);
        if (!InGamut(lin))
        {
            double lo = 0, hi = chroma;
            for (var i = 0; i < 20; i++)
            {
                var mid = (lo + hi) / 2;
                if (InGamut(At(mid)))
                {
                    lo = mid;
                }
                else
                {
                    hi = mid;
                }
            }

            lin = At(lo);
        }

        return (ToSrgb(lin.Item1), ToSrgb(lin.Item2), ToSrgb(lin.Item3));
    }
}
